### PR TITLE
CPU based LoS calculations

### PR DESCRIPTION
**Pull request by mbcoder**

### CODE REVIEW

**Review by alan-edi — COMMENTED** [review]
@mbcoder Looks good! Some suggestions to enhance the comments, aiming to make it easier to understand everything the code's doing.

Also, `README.md` says:
> 2. Choose the java-gradle-starter-project directory and click OK.

whereas I think it should be the **line-of-sight** directory?

File: build.gradle [hunk 1]
```diff
@@ -62,5 +62,9 @@ task copyNatives(type: Copy) {
 }
 
 run {
+    doFirst {
+        // sets the API key from the gradle.properties file as a Java system property
+        systemProperty 'apiKey', apiKey
```
[the remark this answers is not in the document]
Comment (mbcoder) [reply build.gradle:67]: Added a description

File: src/main/java/com/esri/samples/los/App.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            // add the map view to the stack pane
+            stackPane.setCenter(mapView);
+            //stackPane.getChildren().addAll(mapView);
+        } catch (Exception e) {
+            // on any error, display the stack trace.
+            e.printStackTrace();
+        }
+    }
+
+    private double pixelSize() {
```
[the remark this answers is not in the document]
Comment (mbcoder) [reply src/main/java/com/esri/samples/los/App.java:148]: Added an explanation in JavaDoc

File: src/main/java/com/esri/samples/los/SurfaceAnalysis.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+import com.esri.arcgisruntime.geometry.GeometryEngine;
+import com.esri.arcgisruntime.geometry.Point;
+import com.esri.arcgisruntime.geometry.PointCollection;
+import com.esri.arcgisruntime.geometry.Polyline;
+import com.esri.arcgisruntime.geometry.SpatialReferences;
+import com.esri.arcgisruntime.mapping.Surface;
+import com.esri.arcgisruntime.mapping.view.Graphic;
+import com.esri.arcgisruntime.mapping.view.GraphicsOverlay;
+import com.esri.arcgisruntime.symbology.SimpleLineSymbol;
+import com.esri.arcgisruntime.symbology.SimpleMarkerSymbol;
```
[the remark this answers is not in the document]
Comment (mbcoder) [reply src/main/java/com/esri/samples/los/SurfaceAnalysis.java:28]: Cleaned up

File: src/main/java/com/esri/samples/los/SurfaceAnalysis.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+      // if sample angle above observer angle los false, record the split point
+      if (sampleAngle > observerAngle) {
+        losResult = false;
+        splitPoint = samplePoint;
+        break;
+      }
+      distanceFromStart += sampleDistance;
+    }
+
+    // green line of los = true
```
[the remark this answers is not in the document]
Comment (mbcoder) [reply src/main/java/com/esri/samples/los/SurfaceAnalysis.java:99]: I've used these words...

File: src/main/java/com/esri/samples/los/App.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+                    System.out.println("LoS time taken " + (endTime - startTime));
+
+                    // make a polyline for an elevation profile
+                    ArrayList<Point> points = new ArrayList<Point>();
+                    points.add(start2D);
+                    points.add(geoPoint2D);
+                    PointCollection pointCollection = new PointCollection(points);
+                    Polyline elevationLine = new Polyline(pointCollection);
+
+                    // call the elevation profile
```
[the remark this answers is not in the document]
Comment (mbcoder) [reply src/main/java/com/esri/samples/los/App.java:133]: I've removed the elevation profile for now.

**Comment by mbcoder** [comment]
Thanks @alan-edi I've made those suggested changes.  Okay to merge now?

File: src/main/java/com/esri/samples/los/SurfaceAnalysis.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+/**
+ * A class containing static methods for performing 3D surface analysis operations.
+ */
+public class SurfaceAnalysis {
+
+  /**
+   * A method for performing a line of sight calculation using a surface
+   *
+   * @param start starting point for the line of sight calculation
+   * @param end endind point for the  line of sight calculation
```
[the remark this answers is not in the document]
Comment (alan-edi) [reply src/main/java/com/esri/samples/los/SurfaceAnalysis.java:41]: The typo's still there

File: src/main/java/com/esri/samples/los/App.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            RasterLayer rasterLayer = new RasterLayer(rasterforDisplay);
+            map.getOperationalLayers().add(rasterLayer);
+
+            // graphics overlay for showing stuff
+            var graphicsOverlay = new GraphicsOverlay();
+            mapView.getGraphicsOverlays().add(graphicsOverlay);
+
+            // LOS start point 1 metre above surface
+            Point start2D = new Point(-589222.804907, 7468377.448381, SpatialReferences.getWebMercator());
+            Point start3D = new Point(start2D.getX(), start2D.getY(), SurfaceAnalysis.elevationAtPoint(start2D, surface), SpatialReferences.getWebMercator());
```
Comment (alan-edi) [inline src/main/java/com/esri/samples/los/App.java:101]: Split this very long line?

File: src/main/java/com/esri/samples/los/App.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+            mapView.setOnMouseClicked(e -> {
+                if (e.getButton() == MouseButton.PRIMARY && e.isStillSincePress()) {
+                    // create a point from location clicked
+                    Point2D mapViewPoint = new Point2D(e.getX(), e.getY());
+
+                    Point geoPoint2D = mapView.screenToLocation(mapViewPoint);
+
+                    // calculate 3d point slightly above surface (1 metre)
+                    Point geoPoint3D = new Point(geoPoint2D.getX(), geoPoint2D.getY(), SurfaceAnalysis.elevationAtPoint(geoPoint2D, surface)+ 1, SpatialReferences.getWebMercator());
```
Comment (alan-edi) [inline src/main/java/com/esri/samples/los/App.java:113]: Split this very long line?

File: src/main/java/com/esri/samples/los/SurfaceAnalysis.java [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+  /**
+   * A method which returns an elevation profile as a list of heights along a polyline.
+   *
+   * @param line the line used to work out the elevations
+   * @param sampleDistance the distance sampled along the line
+   * @param surface the 3D surface used to get the elevations
+   * @return list of heights along the line
+   */
+  public static ArrayList<Double> elevationProfile (Polyline line, double sampleDistance, Surface surface) {
```
Comment (alan-edi) [inline src/main/java/com/esri/samples/los/SurfaceAnalysis.java:177]: Maybe delete this method that's not used now?

**Review by alan-edi — APPROVED** [review]
@mbcoder Looks food. A few minor comment to consider if you wish.

File: src/main/java/com/esri/samples/los/SurfaceAnalysis.java (hunk 9 above)
Comment (mbcoder) [reply src/main/java/com/esri/samples/los/SurfaceAnalysis.java:177]: I'm keeping this for future use.

File: src/main/java/com/esri/samples/los/App.java (hunk 8 above)
Comment (mbcoder) [reply src/main/java/com/esri/samples/los/App.java]: done

File: src/main/java/com/esri/samples/los/App.java (hunk 7 above)
Comment (mbcoder) [reply src/main/java/com/esri/samples/los/App.java]: Fixed typo and split up long line

**Pull request merged**